Walking skeleton section 51 (#61)
SW709E19/Report: report/PartDesign/ChapterPrototype/Prototype.tex

This chapter elaborates on the initial prototype created during the preliminary part of the design phase. 

> The prototype was conducted as part of the elicitation of requirements and were to be used as the starting point for further agile development.

Reviewer: Kan en prototype blive conducted?

Author: Fixed

> The prototype was conducted as part of the elicitation of requirements and were to be used as the starting point for further agile development.

Reviewer 2: Vi talte også om at referere til vores sprint logs. Det ville være passende at referere til sprint log 1 her, og måske skrive at prototypen er baseret på en række user stories som bliver beskrevet dernede og sårn? Hvad synes I?

> The prototype was conducted as part of the elicitation of requirements and were to be used as the starting point for further agile development.

Author: Fixed in next version

> The prototype was conducted as part of the elicitation of requirements and were to be used as the starting point for further agile development.

Author: Walking skeleton er vel ikke et rigtigt term.
Kan godt prøve at flytte results.
Jeg kan godt referere til sprint logs.

> The prototype was conducted as part of the elicitation of requirements and were to be used as the starting point for further agile development.

Reviewer 2: Jo, jeg tror at det er et rigtigt term.

\section{Overview}

> The prototype consisted of three parts a sensor device, a gateway, and a cloud environment.

Reviewer: ```suggestion
The prototype consisted of three parts; a sensor device, a gateway, and a cloud environment.
```

Author: Fixed

> The prototype consisted of three parts a sensor device, a gateway, and a cloud environment.

Author: Fixed in next version

> The overall overview can be seen in \autoref{fig:prototypeoverview}. 

Reviewer: ```suggestion
An overview can be seen in \autoref{fig:prototypeoverview}. 
```
Kan man lave et overview ude at det også er overall?

Author: Ændret efter Davids forslag

> The overall overview can be seen in \autoref{fig:prototypeoverview}. 

Reviewer 2: ```suggestion
The architecture of the prototype can be seen in \autoref{fig:prototypeoverview}. 
```

> The overall overview can be seen in \autoref{fig:prototypeoverview}. 

Author: Fixed in next version

\begin{figure}[H]
	\centering
	\includegraphics[width=1\textwidth]{PartDesign/ChapterPrototype/images/Overview.png}

> 	\caption{Overview over the prototype.}

Reviewer 2: ```suggestion
	\caption{Overview of the prototype.}
```

	\label{fig:prototypeoverview}
\end{figure}

\subsection{Sensor device}
The sensor device consists of a ESP32 microcontroller \cite{ESP32} to which a sensor is connected. 

> The sensor device measures 7 things, the acceleration of the sensor in 3-axis, the gyro in 3-axis, and the temperature. 

Author: Fixed

These values are transferred from the sensor to the ESP32 digitally through cables.

> The ESP32 then uses it on-board Wi-Fi to publish the 7 measurements as a string to the gateway using the protocol MQTT \cite{MQTT}.

Reviewer: ```suggestion
The ESP32 then uses its on-board Wi-Fi to publish the 7 measurements as a string to the gateway using the MQTT protocol \cite{MQTT}.
```

Author: Fixed in next version

> The ESP32 then uses it on-board Wi-Fi to publish the 7 measurements as a string to the gateway using the protocol MQTT \cite{MQTT}.

Author: Fixed

\subsection{Gateway}

> The gateway consists of a RaspberryPi 3 Model B+ \cite{RaspberryPi3} which serves a local Wi-Fi hotspot such that the sensor devices can be connected and their data be handled.

Reviewer: ```suggestion
The gateway consists of a Raspberry Pi 3 Model B+ \cite{RaspberryPi3} which serves as a local Wi-Fi hotspot such that the sensor devices can be connected and their data transferred.
```

Author: Fixed in next version

> The gateway is for the prototype only responsible for passing on the devices data to the cloud. 

Reviewer 2: ```suggestion
The gateway is only responsible for passing on the devices' data to the cloud in this prototype. 
```

The communication between sensor devices, gateway, and cloud was managed by using Mosquitto \cite{Mosquitto}, a message broker for the MQTT protocol.

\subsection{Cloud architecture}

> The main concern of the cloud computing utilized for the prototype were to expose an endpoint in which gateways could publish data for persistent storage.

Reviewer 2: ```suggestion
The main concern of the cloud computing utilised for the prototype was to expose an endpoint in which gateways could publish data for persistent storage.
```

Author: Fixed

> In order to reach the goal of the prototype within the given timespan, some of the managed services provided by the Google Cloud Platform were used.

Reviewer 2: Jeg tror ikke at vi har beskrevet målet med prototypen endnu.

> In order to reach the goal of the prototype within the given timespan, some of the managed services provided by the Google Cloud Platform were used.

Reviewer 2: Jeg synes heller ikke at vi skal nævne noget timespan. Så er det bedre enten at skrive at det var for ikke at bruge for meget tid på prototypen, eller fordi det ikke ville kontributere til målet med prototypen (det sidste er nok bedst).

Author: Kom med et forslag.

An overview of the services can be seen in \autoref{fig:cloudarchitecture}. 

\begin{figure}[H]
	\centering
	\includegraphics[width=1\textwidth]{PartDesign/ChapterPrototype/images/Cloud.png}
	\caption{The implemented cloud architecture.}
	\label{fig:cloudarchitecture}
\end{figure}

> The first service IoT Core \cite{IoT-Core} is a fully managed platform for managing IoT devices. 

Reviewer 2: ```suggestion
The first service, IoT Core \cite{IoT-Core}, is a fully managed platform for managing IoT devices. 
```

Author: Fixed

The service provides multiple useful features such as device management through registration and authentication as well as two-way secure communication between cloud and devices.

> The service natively support MQTT, with made it relatively straightforward to set up communication with the gateways.

Reviewer 2: ```suggestion
The service natively supports MQTT, which made it relatively straightforward to set up communication with the gateways.
```

Reviewer 2: Burde "gateways" være i ental her? Det lyder til at vi omtaler denne prototype, og der er jo kun én gateway. Selvom jeg godt forstår at der i princippet kunne være flere.

Hvis det bør være "gateway", så kig lige resten af dokumentet igennem for samme rettelse.

Author: Fixed

> The messages received from the gateways are automatically transferred to another GCP service Pub/Sub.

Reviewer: Hvad står GCP for?

Author: Google Cloud Platform (GCP). Jeg tilføjer lige forkortelsen første gang det bliver nævnt

> The messages received from the gateways are automatically transferred to another GCP service Pub/Sub.

Reviewer 2: Forkotelsen "GCP" er ikke defineret nogle steder.

Author: Fixed

> The messages received from the gateways are automatically transferred to another GCP service Pub/Sub.

Reviewer 2: ```suggestion
The messages received from the gateways are automatically transferred to another GCP service: Pub/Sub.
```

Author: Fixed

> The Pub/Sub \cite{PubSub} service is a fully managed event ingestion and delivery system which makes messages available in buckets\footnote{Buckets is a storage component for services on Google Cloud Platform.} for subscribers.

Reviewer 2: ```suggestion
The Pub/Sub \cite{PubSub} service is a fully managed event ingestion and delivery system which makes messages available in buckets\footnote{A bucket is a storage component for services on Google Cloud Platform.} for subscribers.
```

Author: Fixed

> Messages are only temporary stored in the Pub/Sub service, and therefore has to be transferred to persistent storage. 

Reviewer 2: ```suggestion
Messages are only temporarily stored in the Pub/Sub service and therefore have to be transferred to persistent storage. 
```

Author: Fixed

This is achieved through the DataFlow \cite{Dataflow} service, which is a stream and batch data processing service that can be used to transfer and process data from one service to another, and in this case from a Pub/Sub bucket to persistent storage.

> The service utilized for persistent storage DataStore \cite{Datastore} is a fully managed NoSQL-database which is highly scaleble for large amounts of writes and provided the necessary features for the initial requirements of a event log.

Reviewer: Jeg er ikke helt sikker på formuleringen af denne sætning.

Author: Jeg kan godt prøve at omformulere den

> The service utilized for persistent storage DataStore \cite{Datastore} is a fully managed NoSQL-database which is highly scaleble for large amounts of writes and provided the necessary features for the initial requirements of a event log.

Reviewer 2: ```suggestion
The service utilised for persistent storage, DataStore \cite{Datastore}, is a fully managed NoSQL-database which is highly scalable for large amounts of writes and provided the necessary features for the initial requirements of an event log.
```

Author: Fixed

> The service utilized for persistent storage DataStore \cite{Datastore} is a fully managed NoSQL-database which is highly scaleble for large amounts of writes and provided the necessary features for the initial requirements of a event log.

Reviewer 2: Det med event log er måske svært at fatte for læseren, da vi vel ikke har nævnt eller forklaret den før?

Author: Jeg har lavet en footnote der måske forklare det.

> \section{Results}

Reviewer 2: Jeg synes som sagt at dette burde omformuleres og rykkes op i introen

Author: Jeg har rykket det, men ved ikke hvad og hvordan det skal omformuleres

The goal of the prototype was to extract data from a single sensor and transfer the data from a remote location to a central place i.e. the cloud.

> In order to do so each of the components of the system had to be developed to a functioning state, which resulted in the need to mostly select easy and fast solutions, based on the limited timeframe.

Reviewer: ```suggestion
In order to do so each of the components of the system had to be developed to a functioning state, which resulted in the need to mostly select easy and fast solutions, because of the limited timeframe.
```

Author: Fixed in next version

This chapter elaborates on the initial prototype created during the preliminary part of the design phase. 

> The prototype was developed as part of the elicitation of requirements and was to be used as the starting point for further agile development. 

Reviewer 2: ```suggestion
The prototype is a Walking Skeleton implementation of the system, functioning as a minimalistic, end-to-end version of the solution.
It was developed in order to identify and tackle problems at an early stage in the project as well as to lay the foundation for the final solution.
```

> The prototype was the sprint goal of sprint 1 and the user stories that guided the development can be found in \autoref{sprint-log-1}. \todo{fix ref when possible}

Reviewer 2: ```suggestion
The prototype was the sprint goal of sprint 1 and the user stories which guided the development are presented in \autoref{sprint-log-1}. \todo{fix ref when possible}
```

> The goal of the prototype was to extract data from a single sensor and transfer the data from a remote location to a central place i.e. the cloud.

Reviewer 2: ```suggestion
The goal of the prototype was to extract data from a single sensor, send it wirelessly to a gateway and transfer it from the gateway to a central place, i.e. the cloud.
```

> In order to do so each of the components of the system had to be developed to a functioning state, which resulted in the need to mostly select easy and fast solutions, because of the limited timeframe.

Reviewer 2: ```suggestion
To do so, each of the components of the system had to be developed to a functioning state without spending time polishing each component.
```

> This meant that aspects such as configuration of multiple devices and optimizations of power usage was not of concern doing the development of the prototype.

Reviewer 2: ```suggestion
This meant that aspects such as configuration of multiple devices and optimisations of power usage was not of concern doing the development of the prototype.
```

> The prototype however confirmed the initial assumptions made for the design of the system as well as elicited risks and issues that otherwise may not have been brought about.

Reviewer 2: ```suggestion

```

This chapter elaborates on the initial prototype created during the preliminary part of the design phase. 
The prototype is a Walking Skeleton implementation of the system, functioning as a minimalistic, end-to-end version of the solution.
It was developed in order to identify and tackle problems at an early stage in the project as well as to lay the foundation for the final solution.
The prototype was the sprint goal of sprint 1 and the user stories which guided the development are presented in \autoref{sprint-log-1}. \todo{fix ref when possible}

The goal of the prototype was to extract data from a single sensor, send it wirelessly to a gateway and transfer it from the gateway to a central place, i.e. the cloud.
To do so, each of the components of the system had to be developed to a functioning state without spending time polishing each component.

> This meant that aspects such as configuration of multiple devices and optimisations of power usage was not of concern doing the development of the prototype.

Reviewer 2: doing -> during         (jeg kan ikke få suggestion til at virke)
